partiql-ast/src/main/java/org/partiql/ast/v1/AstVisitor.java

```diff
@@ -0,0 +1,240 @@
+package org.partiql.ast.v1;
+
+import org.partiql.ast.v1.expr.Expr;
+import org.partiql.ast.v1.expr.ExprAnd;
+import org.partiql.ast.v1.expr.ExprArray;
+import org.partiql.ast.v1.expr.ExprBag;
+import org.partiql.ast.v1.expr.ExprBetween;
+import org.partiql.ast.v1.expr.ExprCall;
+import org.partiql.ast.v1.expr.ExprCase;
+import org.partiql.ast.v1.expr.ExprCast;
+import org.partiql.ast.v1.expr.ExprCoalesce;
+import org.partiql.ast.v1.expr.ExprExtract;
+import org.partiql.ast.v1.expr.ExprInCollection;
+import org.partiql.ast.v1.expr.ExprIsType;
+import org.partiql.ast.v1.expr.ExprLike;
+import org.partiql.ast.v1.expr.ExprLit;
+import org.partiql.ast.v1.expr.ExprMatch;
+import org.partiql.ast.v1.expr.ExprNot;
+import org.partiql.ast.v1.expr.ExprNullIf;
+import org.partiql.ast.v1.expr.ExprOperator;
+import org.partiql.ast.v1.expr.ExprOr;
+import org.partiql.ast.v1.expr.ExprOverlay;
+import org.partiql.ast.v1.expr.ExprParameter;
+import org.partiql.ast.v1.expr.ExprPath;
+import org.partiql.ast.v1.expr.PathStep;
+import org.partiql.ast.v1.expr.ExprPosition;
+import org.partiql.ast.v1.expr.ExprQuerySet;
+import org.partiql.ast.v1.expr.ExprSessionAttribute;
+import org.partiql.ast.v1.expr.ExprStruct;
+import org.partiql.ast.v1.expr.ExprSubstring;
+import org.partiql.ast.v1.expr.ExprTrim;
+import org.partiql.ast.v1.expr.ExprValues;
+import org.partiql.ast.v1.expr.ExprVarRef;
+import org.partiql.ast.v1.expr.ExprVariant;
+import org.partiql.ast.v1.expr.ExprWindow;
+import org.partiql.ast.v1.graph.GraphLabel;
+import org.partiql.ast.v1.graph.GraphMatch;
+import org.partiql.ast.v1.graph.GraphPart;
+import org.partiql.ast.v1.graph.GraphPattern;
+import org.partiql.ast.v1.graph.GraphQuantifier;
+import org.partiql.ast.v1.graph.GraphSelector;
+
+public interface AstVisitor<R, C> {
+    R visit(AstNode node, C ctx);
+
+    R visitStatement(Statement node, C ctx);
+
+    R visitQuery(Query node, C ctx);
+
+    R visitExplain(Explain node, C ctx);
+
+    R visitTarget(Target node, C ctx);
+
+    R visitTargetDomain(Target.Domain node, C ctx);
```

Comment:
I advise keeping the adjectives on target/domain because these are common terms in the SQL spec (>119 occurrences of "target" in different contexts).

And after looking more closely into the EXPLAIN grammar, I don't think target and domain are even necessary. Consider simplifying to.

```
public class Explain extends Statement {
    @NotNull
    public Map<String, PartiQLValue> options;

    @NotNull
    public Statement statement;

    // ....
}
```

Because the EXPLAIN grammar is _very_ simple and doesn't need three separate nodes. Yes, the PartiQLValue will be replaced in the follow-up. The grammar currently only supports identifier/string literals in the explain options, so this AST representation would cover things like numeric literals etc.

Reply:
Yeah, I was just going off the old partiql-ast.ion definition. This can be simplified a bit. I'll follow your suggestion and add a TODO about getting rid of PartiQLValue.